fastcrud/crud/fast_crud.py

```diff
@@ -421,6 +423,51 @@ async def get(
 
         return None
 
+    def _get_pk_dict(self, instance):
+        return {pk.name: getattr(instance, pk.name) for pk in self._primary_keys}
+
+    async def upsert(
+        self,
+        db: AsyncSession,
+        instance: Union[UpdateSchemaType, CreateSchemaType],
+        schema_to_select: Optional[type[BaseModel]] = None,
+        return_as_model: bool = False,
+    ) -> Union[BaseModel, Dict[str, Any], None]:
+        """Update the instance or create it if it doesn't exists.
+
+        Args:
+            db (AsyncSession): The database session to use for the operation.
+            instance (Union[UpdateSchemaType, type[BaseModel]]): A Pydantic schema representing the instance.
+            schema_to_select (Optional[type[BaseModel]], optional): Optional Pydantic schema for selecting specific columns. Defaults to None.
+            return_as_model (bool, optional): If True, converts the fetched data to Pydantic models based on schema_to_select. Defaults to False.
+
+        Returns:
+            BaseModel: the created or updated instance
+        """
+        _pks = self._get_pk_dict(instance)
+        schema_to_select = schema_to_select or type(instance)
+        db_instance = await self.get(
+            db,
+            schema_to_select=schema_to_select,
+            return_as_model=return_as_model,
+            **_pks,
+        )
+        if db_instance is None:
```

Comment:
Wouldn't it be better to leverage native SQLAlchemy 2 [upsert method](https://docs.sqlalchemy.org/en/20/dialects/postgresql.html#postgresql-insert-on-conflict)?

In fact ["SQLAlchemy does not yet have a backend-agnostic upsert construct"](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-queryguide-upsert) so current apporach looks somewhat justified, but we might consider narrowing this case to just Postgres or maybe providing separate handlers for each supported db.

That's just a set of ideas, I still consider this PR a nice improvement reducing boilerplate code in Fastapi as you stated in the description. We could as well wait for the mentioned backend-agnostic construct in SQLAlchemy.

I'm also wonderring if we'd like to include the method somewhere in the EndpointCreator? Maybe in PUT method?

Reply:
@dubusster @igorbenav any thoughts gentlemen?

Reply:
I think this approach will work no matter the backend since we're manually checking whether the row exists, then updating if it does, creating if it doesn't. The steps are managed in python, I don't see why there would be an issue, the db will never actually get to the `on conflict` part.

Reply:
If we use the native `upsert` on the other hand, we'll definitely have to handle this backend part. We might get more performance, but we'll definitely need to manage a lot more complexity.